addon/services/img-lazy-observer.js

```diff
@@ -1,93 +1,86 @@
       'IntersectionObserver' in window
 
-    if (this.hasIntersectionObserver) {
-      this.components = []
-      this.config = assign({},
-        DEFAULT_OBSERVER_CONFIG,
-        get(this, 'config.ember-img-lazy.observerConfig') || {}
-      )
-      this.onIntersection = run.bind(this, this.onIntersection)
-    }
+    this._config = assign({},
+      DEFAULT_OBSERVER_CONFIG,
+      get(this, 'config.ember-img-lazy.observerConfig')
+    )
   },
 
-  createObserver() {
-    if (this.observer) {
-      throw new Error('There is already an intersection observer present')
-    }
+  // maintains singleton this._observer object
+  getObserver() {
+    if (this._observer) {
+      return this._observer;
+    } else if (this.hasIntersectionObserver) {
+      this._observer = new IntersectionObserver(
+        run.bind(this, this.onIntersection), this._config
+      )
 
-    this.observer = new IntersectionObserver(this.onIntersection, this.config)
+      return this._observer
+    }
   },
 
   stopObserver() {
-    this.observer.disconnect()
-    this.observer = null
-  },
-
-  observe(component) {
-    if (this.hasIntersectionObserver === false) {
-      return
-    }
-
-    if (this.observer === null) {
-      this.createObserver()
+    if (this._observer) {
+      this._observer.disconnect()
+      this._observer = null
     }
-
-    this.observer.observe(component.element)
-    this.components.push(component)
   },
 
-  unobserve(component) {
-    if (this.hasIntersectionObserver === false || this.observer === null) {
-      return
-    }
+  observe(element) {
+    let observer = this.getObserver()
 
-    try {
-      this.observer.unobserve(component.element)
+    if (observer) {
+      observer.observe(element)
     }
-    catch (e) {
-      console.error(e)
-    }
-
-    this.components = this.components.filter(c => c !== component)
+  },
 
-    if (this.components.length === 0) {
-      this.stopObserver()
+  unobserve(element) {
+    if (this._observer) {
+      // calling unobserve() for unobserved elements may throw error
+      // on some browsers especially Edge
+      try {
+        this._observer.unobserve(element)
+      }
+      catch (e) {
+        console.warn(e)
+      }
     }
   },
 
   onIntersection(entries) {
-    for (let i = 0; i < entries.length; i++) {
-      if (entries[i].intersectionRatio > 0) {
-        let component = this.getComponent(entries[i].target)
+    entries.forEach(entry => {
+      if (entry.intersectionRatio > 0) {
+        let component = this.getComponent(entry.target.id)
 
         if (component !== undefined) {
-          this.unobserve(component)
+          this.unobserve(component.element)
           component.loadImage()
         }
       }
-    }
+    });
   },
 
-  getComponent(img) {
-    return this.components.find(c => c.element === img)
-  },
+  getComponent(id) {
+    return getOwner(this).lookup('-view-registry:main')[id]
```

Comment:
As mentioned in the comments, I'd rather not use Ember private APIs, the one array in this service should not have any impact on memory.

Reply:
This array will have to contain all the unvisited `img-lazy` components as the user clicks through pages. The array will swell in memory usage as a user visits a lot of pages.

If Ember removes this private API, we can go back to own implementation. If they change the lookup name, we can update as well.